ibis/expr/operations/arrays.py

```diff
@@ -94,30 +91,28 @@ class ArrayRepeat(Value):
 @public
 class ArrayMap(Value):
     arg = rlz.array
-    # TODO(kszucs): add a callable_with validator to support callable arguments
-    # and return type, e.g. Callable[[ops.Argument], ops.Value] in this case
-    func = rlz.instance_of(Callable)
+    func = rlz.callable_with([rlz.expr_of(rlz.any)], rlz.any)
 
-    output_shape = rlz.shape_like("arg")
+    @attribute.default
+    def parameter(self):
+        (name,) = self.func.__signature__.parameters.keys()
+        return name
 
     @attribute.default
     def result(self):
-        arg = self.arg
-        shape = arg.output_shape
-        dtype = arg.output_dtype.value_type
-        args = [
-            Argument(name=name, shape=shape, dtype=dtype).to_expr()
-            for name in self.signature
-        ]
-        expr = self.func(*args)
-        return expr.op()
-
-    @property
-    def signature(self):
-        return list(inspect.signature(self.func).parameters.keys())
+        arg = Argument(
+            name=self.parameter,
+            shape=rlz.Shape.COLUMNAR,
```

Comment:
This isn't correct. The output shape depends on the input shape. Consider the case of a literal array.

Reply:
Updating.